pyccel/codegen/pipeline.py

```diff
@@ -325,20 +330,34 @@ def handle_error(stage):
                     # remove library folder to avoid missing files and copy
                     # new one from pyccel stdlib
                     lib_dest_path = os.path.join(pyccel_dirpath, lib_name)
-                    if os.path.exists(lib_dest_path):
-                        shutil.rmtree(lib_dest_path)
-                    shutil.copytree(lib_path, lib_dest_path)
+                    if not os.path.exists(lib_dest_path):
+                        try:
+                            shutil.copytree(lib_path, lib_dest_path)
+                        except FileExistsError:
+                            pass
 
                     # stop after copying lib to __pyccel__ directory for
                     # convert only
                     if convert_only:
                         continue
 
                     # get library source files
-                    source_files = []
-                    for e in os.listdir(lib_dest_path):
-                        if e.endswith(lang_ext_dict[language]):
-                            source_files.append(os.path.join(lib_dest_path, e))
+                    ext = lang_ext_dict[language]
+                    obj_ext = lang_obj_ext_dict[language]
+                    source_files = [os.path.join(lib_dest_path, e) for e in os.listdir(lib_dest_path)
+                                                                if e.endswith(ext)]
+
+                    # Add internal lib to internal_libs_name set
+                    internal_libs_name.add(lib)
+                    # add source file without extension to internal_libs_files
+                    internal_libs_files.extend(os.path.splitext(f)[0] for f in source_files)
+                    # add library path to internal_libs_path
+                    internal_libs_path.append(lib_dest_path)
+
+                    # stop if object files already exist
+                    object_files = [s.replace(ext, obj_ext) for s in source_files]
+                    if all(os.path.exists(o) for o in object_files):
+                        continue
```

Comment:
Can you please explain these changes?

Reply:
Identifying that the files are at the correct location is not enough to know if they have been compiled as they may have been copied in `convert_only` mode here I try to identify if the code has already been compiled, and exit if this is the case. This is done to avoid problems between threads. E.g. if thread 1 tries to link using these files while thread 2 is trying to recompile them

Reply:
Thanks, I think I've understood now. Is this correct?
- if all object files for one internal library already exist, you move to the next library
- if one object file for that library is missing, you (re)compile all source files.

I cannot convince myself that this approach is completely water-tight, though. Here we have multiple processes that are compiling libraries, potentially at the same time, without talking to each other. So we have no barriers, and no synchronization whatsoever. I hope that the operating system is able to handle multiple read/write operations properly, but I suspect that we need to use a file-locking mechanism.

Reply:
Yes that's correct. I agree I have seen it occasionally cause problems but I'm not sure how to do better

Reply:
Should we try this out? https://pypi.org/project/filelock

Reply:
See 57be261

Reply:
Done this is now working

Reply:
Nice!